### DISCUSSION TITLE
User-defined invokable type?

### DISCUSSION THREAD

**Discussion by LEI-Hongfaan** [dotnet/csharplang#4004] [Language Ideas] [answered]
Can we make any type that has an invokable member `Invoke` invokable?
So we can write `f(x)` instead of `f.Invoke(x)`.

Alternatively, `operator ()`?

**Comment by YairHalberstadt** [comment]
Do you have any particular use cases in mind? For some inspiration take a look at scala's `apply` feature, where any type with an `apply` method can be treated as a function object.

I think it reasonably likely that there'll be some work in the future to extend the set of things that are delegate like, so that lambdas can be converted to structs, and closures to ref structs - see https://github.com/dotnet/csharplang/discussions/2482 for the motivation. If that ever does happen then it's likely that this might fall out of that work.

**Comment by svick** [comment, marked as answer]
This is a duplicate of https://github.com/dotnet/csharplang/discussions/95.
